driver/utils/type_info.h

```diff
@@ -0,0 +1,2112 @@
+    : public std::false_type
+{
+};
+
+template <> struct is_string_data_source_type<DataSourceType<DataSourceTypeId::String>>
+    : public std::true_type
+{
+};
+
+template <> struct is_string_data_source_type<DataSourceType<DataSourceTypeId::FixedString>>
+    : public std::true_type
+{
+};
+
+template <class T> inline constexpr bool is_string_data_source_type_v = is_string_data_source_type<T>::value;
+
+// Used to avoid duplicate specializations in platforms where 'std::int32_t' or 'std::int64_t' are typedef'd as 'long'.
+struct long_if_not_typedefed {
+    struct dummy {};
+    using type = std::conditional_t<
+        (std::is_same_v<long, std::int32_t> || std::is_same_v<long, std::int64_t>),
+        dummy,
+        long
+    >;
+};
+
+// Used to avoid duplicate specializations in platforms where 'std::uint32_t' or 'std::uint64_t' are typedef'd as 'unsigned long'.
+struct unsigned_long_if_not_typedefed {
+    struct dummy {};
+    using type = std::conditional_t<
+        (std::is_same_v<unsigned long, std::uint32_t> || std::is_same_v<unsigned long, std::uint64_t>),
+        dummy,
+        unsigned long
+    >;
+};
+
+namespace value_manip {
+
+    template <typename T>
+    inline std::int16_t getColumnSize(const T & obj, const TypeInfo & type_info) {
+        return type_info.column_size;
+    }
+
+    inline std::int16_t getColumnSize(const SQL_NUMERIC_STRUCT & numeric, const TypeInfo & type_info) {
+        return (numeric.precision == 0 ? type_info.column_size : numeric.precision);
+    }
+
+    // TODO: implement getColumnSize() for other types.
+
+    template <typename T>
+    inline std::int16_t getDecimalDigits(const T & obj, const TypeInfo & type_info) {
+        return 0;
+    }
+
+    inline std::int16_t getDecimalDigits(const SQL_NUMERIC_STRUCT & numeric, const TypeInfo & type_info) {
+        return numeric.scale;
+    }
+
+    // TODO: implement getDecimalDigits() for other types.
+
+    template <typename ProxyType, typename SourceType, typename DestinationType>
+    void convert_via_proxy(const SourceType & src, DestinationType & dest);
+
+    template <typename SourceType>
+    struct from_value {
+        template <typename DestinationType>
+        struct to_value {
+            static inline void convert(const SourceType & src, DestinationType & dest); // Leave unimplemented for general case.
+        };
+    };
+
+    template <>
+    struct from_value<std::string> {
+        using SourceType = std::string;
+
+        template <typename DestinationType>
+        struct to_value {
+            static inline void convert(const SourceType & src, DestinationType & dest) {
+                dest = fromString<DestinationType>(src);
+            }
+        };
+    };
+
+    template <>
+    struct from_value<std::string>::to_value<std::string> {
+        using DestinationType = std::string;
+
+        static inline void convert(const SourceType & src, DestinationType & dest) {
```

Comment:
Does it make sense to add extra overload? :
```
        static inline void convert(SourceType & src, DestinationType & dest) {
            dest = std::move(src);
        }
```

Reply:
This particular conversion is not used on a critical path. Also, managing storage in `convert()`-type of functions proved to be hell.